src/python/pygrackle/utilities/grdata.py

```diff
@@ -0,0 +1,774 @@
+
+logger = logging.getLogger("grdata")
+logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler())
+
+
+class GrdataError(RuntimeError):
+    pass
+
+
+@contextlib.contextmanager
+def _file_openner(f, mode, **kwargs):
+    if isinstance(f, (str, bytes, os.PathLike)):
+        with open(f, mode, **kwargs) as file:
+            yield file
+    else:
+        yield f
+
+
+def _progress_bar(total_bytes):
+    """provides a function for drawing/updating progress bars"""
+    ncols = shutil.get_terminal_size()[0] - 1
+    power_div_3 = int(log10(total_bytes) // 3) if total_bytes > 0 else 0
+    factor, unit = 1000.0**power_div_3, (" B", "KB", "MB", "GB")[power_div_3]
+    # the output line has the form: '[<progress-bar>] <size>/<size> <unit>'
+    fmt = "\r[{bar:{barlen}.{nfill}}] {size:.2f}" + f"/{total_bytes/factor:.2f} {unit}"
+    barlen = ncols - 19  # for context, 15 <= (len(fmt.format(...)) - barlen) <= 19
+    bar = (barlen * "=") if (barlen >= 1) and sys.stdout.isatty() else None
+
+    def _update(size):
+        nonlocal bar
+        if size is None and bar is not None:
+            print(flush=True)
+            bar = None
+        elif bar is not None:
+            nfill = barlen * int(size / total_bytes)
+            val = fmt.format(bar=bar, barlen=barlen, nfill=nfill, size=size / factor)
+            print(val, end="", flush=True)
+
+    return _update
+
+
+def _retrieve_url(url, dst, *, chunksize=_CHUNKSIZE):
+    """download the file from url to dst"""
+    # Online discussion about calling `response.read(chunksize)`, where `response` is a
+    # context manager object from `url.request.urlopen`, implies that there is an upper
+    # limit on the amount of data read from the http request into memory at a given
+    # point in time. This is unlikely to ever be a problem (the biggest file we
+    # currently fetch is under 10 Megabytes). However, if it does become a problem, we
+    # could just wrap the ``curl``/``wget`` cli tools.
+    try:
+        req = urllib.request.Request(url)
+        with contextlib.ExitStack() as stack:
+            out_file = stack.enter_context(open(dst, "wb"))
+            response = stack.enter_context(urllib.request.urlopen(req))
+            total_bytes = int(response.headers.get("Content-Length", -1))
+            update_progress = _progress_bar(total_bytes)
+            stack.callback(update_progress, size=None)
+
+            # write downloaded data to a file
+            downloaded_bytes = 0
+            while True:
+                update_progress(downloaded_bytes)
+                block = response.read(chunksize)
+                if not block:
+                    break
+                downloaded_bytes += len(block)
+                out_file.write(block)
+    except HTTPError as e:
+        raise GrdataError(f"server can't fulfill request to fetch {url}: {e.code}")
+    except URLError as e:
+        raise GrdataError(f"server can't be reached to fetch {url}: {e.code}")
```

Comment:
This could theoretically be replaced by `pooch`.

Alternatively, we could stop showing a progress-bar (and stop repackaging the errors). Alternatively we could use something like:

```python
def download_file(url, dst, quiet_download=False):
    """download the file from url to dst"""
    choices = [
        ("wget", "--output-document", "--quiet"), ("curl", "--output", "--silent")
    ]
    for tool, outflag, quietflag in choices:
        if shutil.which(tool) is None:
            tmp = [quietflag] if quiet_download else []
            _cmd = [tool, outflag, dst] + tmp + [url]
            subprocess.check_output(_cmd)
            return
    raise RuntimeError("neither wget nor curl is installed")
```